src/modules/fw_pos_control/performance_model/PerformanceModel.cpp

```diff
@@ -0,0 +1,211 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+/**
+ * @file PerformanceModel.cpp
+ * Performance model.
+ */
+
+#include <geo/geo.h>
+#include <px4_platform_common/events.h>
+#include "PerformanceModel.h"
+
+// air density of standard athmosphere at 5000m above mean sea level [kg/m^3]
+static constexpr float AIR_DENSITY_STANDARD_ATMOS_5000_AMSL = 0.7363f;
+
+// air density of standard athmosphere at 1000m above mean sea level [kg/m^3]
+static constexpr float AIR_DENSITY_STANDARD_ATMOS_1000_AMSL = 1.112f;
+
+// [.] minimum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float MIN_WEIGHT_RATIO = 0.5f;
+
+// [.] maximum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float MAX_WEIGHT_RATIO = 2.0f;
+
+// climbrate defining the service ceiling, used to compensate max climbrate based on air density
+static constexpr float CLIMBRATE_MIN = 0.5f; // [m/s]
+
+PerformanceModel::PerformanceModel(): ModuleParams(nullptr)
+{
+	updateParams();
+}
+float PerformanceModel::getWeightRatio() const
+{
+	float weight_factor = 1.0f;
+
+	if (_param_weight_base.get() > FLT_EPSILON && _param_weight_gross.get() > FLT_EPSILON) {
+		weight_factor = math::constrain(_param_weight_gross.get() / _param_weight_base.get(), MIN_WEIGHT_RATIO,
+						MAX_WEIGHT_RATIO);
+	}
+
+	return weight_factor;
+}
+float PerformanceModel::getMaximumClimbRate(float air_density) const
+{
+
+	float climbrate_max = _param_fw_t_clmb_max.get();
+
+	const float density_min = _param_density_min.get();
+
+	if (density_min < AIR_DENSITY_STANDARD_ATMOS_1000_AMSL
+	    && density_min > AIR_DENSITY_STANDARD_ATMOS_5000_AMSL) {
+		const float density_gradient = (_param_fw_t_clmb_max.get() - CLIMBRATE_MIN) / (CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C -
+					       density_min);
+		const float delta_rho = air_density - CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C;
+		climbrate_max = _param_fw_t_clmb_max.get() + density_gradient * delta_rho;
+	}
+
+	return climbrate_max / getWeightRatio();
+}
+float PerformanceModel::getTrimThrottle(float throttle_min, float throttle_max, float airspeed_sp,
```

Comment:
i made a couple encapsulation requests.. not to be totally pedantic.. but just that it could make this function a little easier to read. sanity checks and business logic contained in their respective silos, and then can just put the pure business on top level -- making super clear, we get trim scheduled per airspeed, then scale it for weight and density, done:
```cpp
const float throttle_trim = trimThrottlePerAirspeed(airspeed_setpoint);
const float throttle_trim_scaled = throttle_trim * sqrtf(getWeightRatio()) * sqrtf(getAirDensityRatio(air_density));
return math::constrain(throttle_trim_scaled, throttle_min, throttle_max);
```

Reply:
@tstastny Yes we can. Until now I just copied over functions from the position controller as a first step.